docs/rfcs/optional-config-type.md

```diff
@@ -0,0 +1,276 @@
+		hss.MaxRequestBodySize,
+		serverOpts.ErrHandler,
+		hss.CompressionAlgorithms,
+		serverOpts.Decoders,
+	)
+
+	// ...
+
+	if hss.Auth.HasValue() {
+		server, err := hss.Auth.Value().GetServerAuthenticator(context.Background(), host.GetExtensions())
+		if err != nil {
+			return nil, err
+		}
+
+		handler = authInterceptor(handler, server, hss.Auth.Value().RequestParameters)
+	}
+
+	corsValue := hss.CORS.Value()
+	if hss.CORS.HasValue() && len(hss.CORS.AllowedOrigins) > 0 {
+		co := cors.Options{
+			AllowedOrigins:   corsValue.AllowedOrigins,
+			AllowCredentials: true,
+			AllowedHeaders:   corsValue.AllowedHeaders,
+			MaxAge:           corsValue.MaxAge,
+		}
+		handler = cors.New(co).Handler(handler)
+	}
+	if hss.CORS.HasValue() && len(hss.CORS.AllowedOrigins) == 0 && len(hss.CORS.AllowedHeaders) > 0 {
+		settings.Logger.Warn("The CORS configuration specifies allowed headers but no allowed origins, and is therefore ignored.")
+	}
+
+	if hss.ResponseHeaders.HasValue() {
+		handler = responseHeadersHandler(handler, hss.ResponseHeaders.Value())
+	}
+
+	// ...
+}
+```
+
+### Proper unmarshaling of empty values when a default is set
+
+Currently, the OTLP receiver requires a workaround to make enabling each
+protocol optional while providing defaults if a key for the protocol has been
+set:
+
+```golang
+type Protocols struct {
+	GRPC *configgrpc.ServerConfig `mapstructure:"grpc"`
+	HTTP *HTTPConfig              `mapstructure:"http"`
+}
+
+// Config defines configuration for OTLP receiver.
+type Config struct {
+	// Protocols is the configuration for the supported protocols, currently gRPC and HTTP (Proto and JSON).
+	Protocols `mapstructure:"protocols"`
+}
+
+func createDefaultConfig() component.Config {
+	return &Config{
+		Protocols: Protocols{
+			GRPC: configgrpc.NewDefaultServerConfig(),
+			HTTP: &HTTPConfig{
+                // ...
+			},
+		},
+	}
+}
+
+func (cfg *Config) Unmarshal(conf *confmap.Conf) error {
+	// first load the config normally
+	err := conf.Unmarshal(cfg)
+	if err != nil {
+		return err
+	}
+
+    // gRPC will be enabled if this line is not run
+	if !conf.IsSet(protoGRPC) {
+        cfg.GRPC = nil
+	}
+
+     // HTTP will be enabled if this line is not run
+	if !conf.IsSet(protoHTTP) {
+		cfg.HTTP = nil
+	}
+
+	return nil
+}
+```
+
+With an Optional type, the checks in `Unmarshal` become unnecessary, and it's
+possible the entire `Unmarshal` function may no longer be needed. Instead, when
+the config is unmarshaled, no value would be put into the default Optional
+values and `HasValue` would return false when using this config object.
+
+This situation is something of an edge case with our current unmarshaling
+facilities and while not common, could be a rough edge for users looking to
+implement similar config structures.
+
+## Disadvantages of an Optional type
+
+There is one noteworthy disadvantage of introducing an Optional type:
+
+1. Since the type isn't standard, external packages working with config may
```

Comment:
What is the intended implementation plan for Optional config types? In the above example, it says:
>This situation is something of an edge case

So is the intention to only modify fields who meet some required use case? Or is it to modify all fields that are currently pointers?

Changing the field type is a breaking API change. Is there a different approach depending on the stability level of a component? Or is the expectation that anyone importing and using a config struct from an otel component will have to make any required changes given the new struct field types?

Reply:
@kristinapathak Do you have examples of components/modules that are 1.x and need to use this optional type? We are blocking `confighttp`/`configgrpc` on this so that we can apply this pattern on these modules.

The intent is to apply this new type everywhere, go through an (API) breaking change, and ensure this type is used everywhere going forward.

Reply:
@kristinapathak does this solve your question?

Reply:
yes, thank you